server/src/main/java/org/elasticsearch/action/bulk/BulkRequest.java

```diff
@@ -427,7 +428,10 @@ public BulkRequest add(BytesReference data, @Nullable String defaultIndex, @Null
                                 }
                                 index = parser.text();
                             } else if (TYPE.match(currentFieldName, parser.getDeprecationHandler())) {   
-                                deprecationLogger.deprecatedAndMaybeLog("bulk_with_types", RestBulkAction.TYPES_DEPRECATION_MESSAGE);
+                                if (typesDeprecationLogged == false) {
+                                    deprecationLogger.deprecatedAndMaybeLog("bulk_with_types", RestBulkAction.TYPES_DEPRECATION_MESSAGE);
+                                    typesDeprecationLogged = true;
+                                }
```

Comment:
Could we add a test that checks that only 1 deprecation warning is returned when the bulk request would previously generate multiple?

Reply:
I think they're thinned out already in DeprecationLogger/ThreadContext? 
It dedups values [added to the response header ](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/common/util/concurrent/ThreadContext.java#L330)
This PR just adds a more performant way of thinning them out.

Reply:
ok, I see. In which case I think we should test to make sure this actually fixes the performance issue before we merge it as I didn't see that we already deduplicate the warning headers. Maybe we can run one of the rally tracks against this PR?

Reply:
I'll ask